File: prototype_customgoal.m
function value = examplePrototypeCustomGoal

opensimroot = 'C:\opensim 4.3\'; %create a char array that has the opensim path toplevel directory
addpath([opensimroot 'bin'], [opensimroot 'sdk\lib']); %add the opensim path to the
javaaddpath([opensimroot 'bin'], [opensimroot 'sdk\lib']); %add opensimroot bin and the java path to MATLAB's dynamic path path
setenv('PATH', [[opensimroot 'bin'] ';' [opensimroot 'sdk\lib'] ';' getenv('PATH')]);% Set Windows System path to include OpenSim libraries
import org.opensim.modeling.* %import opensim api library

study = MocoStudy();
problem = study.updProblem();

modelProcessor = ModelProcessor('./model/model_31d84m_modified_contacts.osim');
modelProcessor.append(ModOpIgnoreTendonCompliance());
%weld some of the joints
jointsToWeld = StdVectorString();
% Weld shoulder and elbow joints for initial torque-driven sim. 
jointsToWeld.add('shoulder_r');
jointsToWeld.add('shoulder_l');
jointsToWeld.add('elbow_r');
jointsToWeld.add('elbow_l');
%Weld arm, hand and foot
jointsToWeld.add('radioulnar_r');
jointsToWeld.add('radioulnar_l');
jointsToWeld.add('wrist_r');
jointsToWeld.add('wrist_l');
jointsToWeld.add('mtp_r');
jointsToWeld.add('mtp_l');
jointsToWeld.add('subtalar_r');
jointsToWeld.add('subtalar_l');
modelProcessor.append(ModOpReplaceJointsWithWelds(jointsToWeld));
model = modelProcessor.process();

% Add metabolic cost model
metabolics = Bhargava2004SmoothedMuscleMetabolics();
metabolics.setName('metabolic_cost');
metabolics.set_use_smoothing(true);

ms = model.getMuscles();
for m = 0: ms.getSize()-1
    this_muscle_name = ms.get(m).getName();
    this_muscle      = ms.get(m);
    metabolics.addMuscle(this_muscle_name, this_muscle);
end

model.addComponent(metabolics);
model.finalizeConnections();

problem.setModel(model);
problem.setTimeBounds(0, 1.16);

solver = study.initCasADiSolver();
solver.resetProblem(problem);
trajectory = solver.createGuess();
lowerOrderSolution = MocoTrajectory('./output/1_00.sto');
lowerOrderStatesTable = lowerOrderSolution.exportToStatesTable();
lowerOrderControlsTable = lowerOrderSolution.exportToControlsTable();
trajectory.insertStatesTrajectory(lowerOrderStatesTable, true);
trajectory.insertControlsTrajectory(lowerOrderControlsTable, true);
n_muscles = 76;
n_coordinates = 21;
n_passive = 14;
totalMass = 75.3370;

w = [((500/n_muscles)/totalMass) 2000/n_muscles 50000/n_coordinates 1000/n_passive];
value = evaluateCustomGoal(problem, trajectory,w, ...
        @calcMyCustomEffortGoalIntegrand, @calcMyCustomEffortGoalValue);

end

function integrand = calcMyCustomEffortGoalIntegrand(model, state,w)

import org.opensim.modeling.* %import opensim api library

% Compute the integrand for the integral portion of the cost goal.

model.realizeAcceleration(state);
met = Bhargava2004SmoothedMuscleMetabolics.safeDownCast(model.getComponent('metabolic_cost'));
ms = model.getMuscles();
n_muscles = ms.getSize();

for m = 0:model.getMuscles.getSize()-1
    this_muscle = model.getMuscles().get(m);
    metabolic_cost(m+1) = met.getMuscleMetabolicRate(state,this_muscle.getAbsolutePathString());
end

% This computes the sum of squared element values (squared 2-norm).
integrand = 0;
for i = 0:n_muscles - 1
    integrand = integrand + w(1)*metabolic_cost(i+1)^2;
end

controls = model.getControls(state);
for i = 0:controls.size() - 1
    integrand = integrand + w(2)*controls.get(i)^2;
end

udots = state.getUDot();
for i = 0:udots.size() - 1
    integrand = integrand + w(3)*udots.get(i)^2;
end

fs = model.getForceSet();
for f = 0:fs.getSize()-1
    this_force = fs.get(f);
    if contains(string(this_force.getName()),'Passive','IgnoreCase',true)
        exp = ExpressionBasedCoordinateForce.safeDownCast(this_force);
        integrand = integrand + w(4)*exp.getForceMagnitude(state)^2;
    end
end
disp(integrand);

end

function value = calcMyCustomEffortGoalValue(...
                        model, initial_state, final_state, integral)
% Compute the goal value from the phase's initial and final states, and from
% the integral of the integrand function over the phase. The integration is
% performed for you by evaluateCustomGoal().
value = integral / (final_state.getTime() - initial_state.getTime());
end

function goalValue = evaluateCustomGoal(...
                        problem, mocoTraj,w, integrandFunc, goalFunc)
% Test a custom goal, defined by integrand and goal functions, on the
% provided problem and MocoTrajectory.
model = problem.getPhase(0).getModelProcessor().process();
org.opensim.modeling.opensimMoco.prescribeControlsToModel(mocoTraj, model);
statesTraj = mocoTraj.exportToStatesTrajectory(problem);
model.initSystem();
N = statesTraj.getSize();
integrand = zeros(N, 1);
for i = 0:(N - 1)
    integrand(i + 1) = integrandFunc(model, statesTraj.get(i),w);
end

integral = trapz(integrand);
goalValue = goalFunc(model, statesTraj.front(), statesTraj.back(), integral);

end

File: sandbox/PrototypeCustomGoal.cpp
#include <OpenSim/Actuators/ActivationCoordinateActuator.h>
#include <OpenSim/Moco/osimMoco.h>

using namespace OpenSim;
using SimTK::Pi;

typedef double (*integrand_func)(const Model&, const SimTK::State&, const double);
typedef double (*goal_func)(const Model&, const SimTK::State&, const SimTK::State&, const double&);
integrand_func integrandFunc;
goal_func goalFunc;

std::unique_ptr<Model> createSlidingMassModel() {
	auto model = make_unique<Model>();
	model->setName("sliding_mass");
	model->set_gravity(SimTK::Vec3(0, 0, 0));
	auto* body = new Body("body", 10.0, SimTK::Vec3(0), SimTK::Inertia(0));
	model->addComponent(body);

	// Allows translation along x.
	auto* joint = new SliderJoint("slider", model->getGround(), *body);
	auto& coord = joint->updCoordinate(SliderJoint::Coord::TranslationX);
	coord.setName("position");
	model->addComponent(joint);

	auto* actu = new CoordinateActuator();
	actu->setCoordinate(&coord);
	actu->setName("actuator");
	actu->setOptimalForce(1);
	model->addComponent(actu);

	body->attachGeometry(new Sphere(0.05));

	model->finalizeConnections();

	return model;
}

double trapz(SimTK::Vector& integrand) {
	int N = integrand.size() - 1;
	double step = (N) / (2.0 * ((double)N));
	SimTK::Vector v(integrand);
	for (int i = 1; i < N; i++) {
		v[i] *= 2.0;
	}
	double integral = v.sum();
	integral *= step;
	return integral;
}


double CustomGoalIntegrand(const Model& model, const SimTK::State& state, double w) {
	model.realizeAcceleration(state);
	double integrand = 0;
	auto udots = state.getUDot();
	for (int i = 0; i < udots.size(); i++) {
		integrand += w * pow(udots.get(i),2);
	}
	return integrand;
}

double CustomGoalValue(const Model& model, const SimTK::State& initial_state, const SimTK::State& final_state, const double& integral) {
	double value = integral / (final_state.getTime() - initial_state.getTime());
	return value;
}

double evaluateCustomGoal(const MocoProblem& problem, const MocoTrajectory& mocoTraj, integrand_func integrandFunc, goal_func goalFunc, const double w) {
	//     // Test a custom goal, defined by integrand and goal functions, on the
	//     // provided problem and MocoTrajectory.
	auto& model = problem.getPhase(0).getModelProcessor().process();
	prescribeControlsToModel(mocoTraj, model);
	auto& statesTraj = mocoTraj.exportToStatesTrajectory(problem);
	model.initSystem();
	int N = (int)statesTraj.getSize();
	SimTK::Vector integrand(N);
	for (int i = 0; i < N; i++) {
		integrand[i] = integrandFunc(model, statesTraj.get(i), w);
	}

	const double& integral = trapz(integrand);
	double goalValue = goalFunc(model, statesTraj.front(), statesTraj.back(), integral);
	return goalValue;
}

int main()
{
	double controlEffortWeight = 10;
	double stateTrackingWeight = 1;
	double GRFTrackingWeight = 1;

	MocoStudy study;
	MocoProblem& problem = study.updProblem();

	problem.setModel(createSlidingMassModel());
	// Bounds.
	// -------
	// Initial time must be 0, final time can be within [0, 1].
	problem.setTimeBounds(MocoInitialBounds(0.), MocoFinalBounds(1));

	// Position must be within [-5, 5] throughout the motion.
	// Initial position must be 0, final position must be 1.

	// Position must be within [-5, 5] throughout the motion.
	// Initial position must be 0, final position must be 1.
	problem.setStateInfo("/slider/position/value", MocoBounds(-1, 1), MocoInitialBounds(0), MocoFinalBounds(1));
	// Speed must be within [-50, 50] throughout the motion.
	// Initial and final speed must be 0. Use compact syntax.
	problem.setStateInfo("/slider/position/speed", 1, 1, 1);

	// Applied force must be between -50 and 50.
	problem.setControlInfo("/actuator", MocoBounds(-50, 50));

	// Solve the problem.
	// -----
	MocoCasADiSolver& solver = study.initCasADiSolver();
	solver.set_num_mesh_intervals(50);
	MocoSolution solution = study.solve();
	//solution.write("point_mass_solution.sto");

	solver = study.initCasADiSolver();
	solver.resetProblem(problem);

	integrandFunc = CustomGoalIntegrand;
	goalFunc = CustomGoalValue;
	std::cout << "Running custom goal on solution: " << std::endl;
	double w = 1.0;
	double myIntegral = evaluateCustomGoal(problem, solution, integrandFunc, goalFunc,w);
	std::cout << "custom goal value: " << myIntegral << std::endl;
	return EXIT_SUCCESS;
}
